README.md
# WPM Monitor

This extension enables a words per minute counter in the status bar of the editor.

![showcase](./img/showcase.gif)

A word is counted as 5 keypresses (excluding space, delete and modifier keys).

## Features

The monitor offers two modes:

- a character based mode: where the wpm is calculated over the last 100 characters entered
- a time based mode: where the wpm is calculated over a fixed timeframe

You can quickly switch modes for the current session or reset &amp; restart the wpm monitor with the command prompt.

You can activate an automatic reset of the wpm if you stop typing for a specified duration.

There's also a special effect that shows in the status bar if you type faster than a specified threshold, which you can also edit via the command prompt.

## Configuration

- set the default mode in which the wpm monitor runs on startup
- set the number of characters that should be considered for calculating the wpm in character based mode
- set the duration of the time frame to be considered for calculating the wpm in time based mode
- set the threshold for displaying the special effect alongside the wpm monitor
- select the special effect symbol that will get displayed
- enable/disable automatic reset of wpm
- set the duration of the time frame after which the wpm will get reset if no new user input was recorded

## Known Issues

- the wpm monitor also gets displayed in the status bar if no text window is open
- the wpm monitor doesn't work in jupyter notebooks that were opened with the notebook editor of the python extension

## Release Notes

Release notes of mayor and minor releases. For a full changelog check [CHANGELOG.md](./CHANGELOG.md).

### 0.4.0

- patched lodash security vulnerability
- removed space key from being counted towards the wpm counter
- added maximum number of inputs to be considered for the wpm calculation (60000 &rarr; 10h@100wpm)

### 0.3.0

- added optional automatic wpm reset after a specified duration of no new user input
- made prompt configurations persisent by converting them into extension configurations

### 0.2.0

- added a time based wpm mode and a command prompt mode switcher
- added configuration options for length of the timeframe and number of characters to be considered for wpm calculation
- added a special effect that is displayed after you reach a wpm threshold
- added option to reset the wpm display with a command or clicking the counter in the status bar

### 0.1.0

- added basic functionality and character based wpm mode


CHANGELOG.md
# Changelog

All notable changes to the `vscode-wpm-monitor` extension will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2020-07-20

### Patched

- patched lodash security vulnerability

### Changed

- removed space key from being counted towards the wpm counter
- added maximum number of inputs to be considered for the wpm calculation (60000 &rarr; 10h@100wpm)

## [0.3.0] - 2020-07-01

### Added

- added optional automatic wpm reset after a specified duration of no new user input
- added persistent configurations, editable in `settings.json`

### Removed

- removed the majority of the temporary configuration prompts (moved configuration to `settings.json`)

## [0.2.1] - 2020-07-01

### Changed

- clarified the description of the extension

## [0.2.0] - 2020-06-22

### Added

- time based wpm mode and a command prompt mode switcher
- configuration options for length of the timeframe and number of characters to be considered for wpm calculation
- special effect that is displayed after you reach a wpm threshold
- option to reset the wpm display with a command or clicking the counter in the status bar

## [0.1.0] - 2020-06-21

### Added

- basic functionality and character based wpm mode


src/test/suite/extension.test.ts
import * as assert from 'assert';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';
// import * as myExtension from '../../extension';

suite('Extension Test Suite', () => {
	vscode.window.showInformationMessage('Start all tests.');

	test('Sample test', () => {
		assert.equal(true, true);
	});
});


src/wpmMonitorStatusBar.ts
import * as vscode from 'vscode';

export class WpmMonitorStatusBar {
	private interval:number = 1000;	// interval (milliseconds) in which the status bar gets updated
	private statusBar: vscode.StatusBarItem;
	private wpm: number = 0;
	private specialEffectThreshold: number = 80;
	private specialEffect: string = "🔥";

	constructor() {
		this.statusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 0);
		this.statusBar.command = "extension.reset";	// reset on click
		this.statusBar.show();

		this.updateUI();
	}

	public setWPM(wpm: number) {
		this.wpm = wpm;
	}

	public getSpecialEffect() {
		return this.specialEffect;
	}
	public setSpecialEffect(specialEffect: string) {
		this.specialEffect = specialEffect;
	}

	public getSpecialEffectThreshold() {
		return this.specialEffectThreshold;
	}
	public setSpecialEffectThreshold(specialEffectThreshold: number) {
		this.specialEffectThreshold = specialEffectThreshold;
	}
	
	private updateUI() {
		return setInterval(()=> {
		this.statusBar.text = `${this.wpm >= this.specialEffectThreshold ? this.specialEffect + " ":""}${this.wpm.toString()} wpm`;
		}, this.interval);
	}

	public dispose() {
		this.statusBar.dispose();
		clearInterval(this.updateUI());
	}
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { WpmMonitorStatusBar } from './wpmMonitorStatusBar';

// CONSTANTS
const wordsLength: number = 5;	// number of characters to be considered as a word
// all available modes
enum Mode {
	character = "WPM per character",
	time = "WPM per time"
}

// CONFIGURATION DEFAULTS (gets overwritten by loadConfiguration())
let config = {
	mode: Mode.character,
	character: 100,
	duration: 60,
	autoReset: {
		duration: 60,
		enabled: false
	},
	specialeffect: {
		symbol: "🔥",
		threshold: 80
	}
};

// VARIABLES
let inputTimestamps: number[] = [];
let autoResetHandler:NodeJS.Timeout;
let autoResetActive:Boolean = true;

// OBJECTS
let documentChangeListenerDisposer: vscode.Disposable;
let wpmMonitor = new WpmMonitorStatusBar();


// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	// console.log('Congratulations, your extension "vscode-wpm-monitor" is now active!');



	// add the status bar widget
	context.subscriptions.push(wpmMonitor);

	documentChangeListenerDisposer = vscode.workspace.onDidChangeTextDocument(onDidChangeTextDocument);
	vscode.workspace.onDidChangeConfiguration(onDidChangeConfiguration);
	// load the configuration of the extension
	loadConfiguration();

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	const modeSwitcherCommand = vscode.commands.registerCommand('extension.modeSwitcher', () => {
		// display mode-switcher
		vscode.window.showQuickPick(Object.values(Mode)).then((m) => {
			switch (m) {
				case Mode.character:
					// wpm over specified number of characters
					config.mode = Mode.character;
					break;
				case Mode.time:
					// wpm over specified time frame
					config.mode = Mode.time;
					break;
				default:
					// don't change anything
					break;
			}
			reset();
		});
	});
	context.subscriptions.push(modeSwitcherCommand);

	// resets the values for wpm calculation
	const resetCommand = vscode.commands.registerCommand('extension.reset', () => {
		reset(true);
	});
	context.subscriptions.push(resetCommand);
}

// polls every second if latest input is still in time frame or not
// polling only active as long as the time frame is active
// -> should get called less often than with event-based active monitoring
function autoReset() {
	return setInterval(() => {
		if (inputTimestamps[inputTimestamps.length - 1] <= Date.now() - 1000 * config.autoReset.duration) {
			// only clear interval in itself
			autoResetActive = false;
			clearInterval(autoResetHandler);
			reset();
		}
	}, 1000);
}


function reset(resetDisplay?:boolean) {
	inputTimestamps = [];
	if (resetDisplay) {
		wpmMonitor.setWPM(0);
	}
}

function loadConfiguration() {
	const c = vscode.workspace.getConfiguration("wpm-monitor");
	config.mode = (c.get("defaultMode") === "time") ? Mode.time : Mode.character;
	config.character = c.get("mode.character.numberOf") || 100;
	config.duration = c.get("mode.time.duration") || 60;
	config.specialeffect.symbol = c.get("specialEffect.symbol") || "🔥";
	config.specialeffect.threshold = c.get("specialEffect.threshold") || 80;
	config.autoReset.enabled = c.get("autoReset.enabled") || false;
	config.autoReset.duration = c.get("autoReset.duration") || 60;

	// set config of the status bar widget
	wpmMonitor.setSpecialEffect(config.specialeffect.symbol);
	wpmMonitor.setSpecialEffectThreshold(config.specialeffect.threshold);

	// start autoReset routine
	if (config.autoReset.enabled === true) {
		autoResetActive = true;
		clearInterval(autoResetHandler);
		autoResetHandler = autoReset();
	}
}

function onDidChangeConfiguration(event: vscode.ConfigurationChangeEvent) {
	if (event.affectsConfiguration("wpm-monitor")) {
		loadConfiguration();
		reset(true);
	}
}



function onDidChangeTextDocument(event: vscode.TextDocumentChangeEvent) {
	let input = event.contentChanges[0].text;	// string of characters entered

	// remove EOL timestamps
	switch (config.mode) {
		case Mode.character:
			// remove the oldest timestamps until threshold is reached (could also be done with if)
			while (inputTimestamps.length >= config.character) {
				inputTimestamps.shift();
			}
			break;

		case Mode.time:
			let threshold = Date.now() - config.duration * 1000;
			// remove all timestamps that are older than the threshold-timestamp (array/list is sorted with oldest first)
			while (inputTimestamps[0] < threshold) {
				inputTimestamps.shift();
			}
			break;
		default:
			break;
	}

	// maximum length of the array (10h@100wpm)
	while (inputTimestamps.length >= 60000) {
		inputTimestamps.shift();
	}

	// skip deletions and (wpm won't get updated)
	if (input === "" || input === " ") {
		return;
	}
	// TODO: skip copy paste items, language specific skip lists, user skip lists

	// restart autoReset if enabled
	// autoResetActive is only false when autoResetHandler cleared itself
	if (config.autoReset.enabled && autoResetActive===false) {
		autoResetActive = true;
		autoResetHandler = autoReset();
	}

	// Add newest input as timestamp to list
	inputTimestamps.push(Date.now());
	// console.log(event.contentChanges[0].text);

	// calculate the wpm
	let wpm = 0;
	if (inputTimestamps.length >= 2) {
		let oldest = inputTimestamps[0];
		let newest = inputTimestamps[inputTimestamps.length - 1];
		let minute = (newest - oldest) / 1000 / 60;
		let words = inputTimestamps.length / wordsLength;
		wpm = Math.floor(words / minute);
	}
	wpmMonitor.setWPM(wpm);
}

// this method is called when your extension is deactivated
export function deactivate() {
	if (wpmMonitor) {
		wpmMonitor.dispose();
	}
	if (documentChangeListenerDisposer) {
		documentChangeListenerDisposer.dispose();
	}
}